File: app/src/main/java/com/caitlynwiley/pettracker/models/TrackerItem.java
package com.caitlynwiley.pettracker.models;

import android.annotation.TargetApi;
import android.content.Context;

import com.caitlynwiley.pettracker.R;
import com.google.firebase.database.Exclude;
import com.google.gson.annotations.SerializedName;

import java.util.Calendar;
import java.util.Locale;

public class TrackerItem {
    @SerializedName("date")
    private String date;

    @Exclude
    private String localTime;

    @SerializedName("item_id")
    private String itemId;

    @SerializedName("item_type")
    private String itemType;

    @SerializedName("month")
    private int month;

    @SerializedName("day")
    private int day;

    @SerializedName("year")
    private int year;

    @SerializedName("utcmillis")
    private long utcMillis;

    @SerializedName("type")
    private EventType type;

    @SerializedName("walk_length")
    private WalkLength walkLength;

    @SerializedName("pet_id")
    private String petId;

    @SerializedName("cups_food")
    private double cupsFood;

    @SerializedName("number_1")
    private boolean number1;

    @SerializedName("number_2")
    private boolean number2;

    @SerializedName("pretty_date")
    private String prettyDate;

    @Exclude
    private Context mContext;

    public TrackerItem(){}

    public String getDate() {
        return date;
    }

    public void setDate(String date) {
        this.date = date;
        month = Integer.parseInt(date.substring(0, 2).trim());
        day = Integer.parseInt(date.substring(3, 5).trim());
        year = Integer.parseInt(date.substring(6));
    }

    public void setUtcMillis(long utcMillis) {
        this.utcMillis = utcMillis;
        setLocalTime(utcMillis);
    }

    @Exclude
    @TargetApi(24)
    private void setLocalTime(long utcMillis) {
        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(utcMillis);
        localTime = String.format(Locale.US, "%d:%02d %s", c.get(Calendar.HOUR) == 0 ? 12 : c.get(Calendar.HOUR), c.get(Calendar.MINUTE),
                c.get(Calendar.AM_PM) == Calendar.AM? "am" : "pm");
    }

    public long getUtcMillis() {
        return utcMillis;
    }

    public String getItemId() {
        return itemId;
    }

    public void setItemId(String itemId) {
        this.itemId = itemId;
    }

    public String getItemType() {
        return itemType;
    }

    public void setItemType(String itemType) {
        this.itemType = itemType;
    }

    public int getDay() {
        return day;
    }

    public void setDay(int day) {
        this.day = day;
    }

    public int getMonth() {
        return month;
    }

    public void setMonth(int month) {
        this.month = month;
    }

    public int getYear() {
        return year;
    }

    public void setYear(int year) {
        this.year = year;
    }

    public void setPetId(String petId) {
        this.petId = petId;
    }

    public String getPetId() {
        return petId;
    }

    public String getLocalTime() {
        return localTime;
    }

    @Exclude
    public int getDrawableResId() {
        if (itemType.equalsIgnoreCase("day")) return -1;
        switch (type) {
            case POTTY:
                return R.drawable.ic_dog_poop_64dp;
            case FEED:
                return R.drawable.ic_dog_bowl_64dp;
            case WALK:
                return R.drawable.ic_dog_walk_64dp;
        }
        return R.drawable.ic_clock_black_24dp; // default for now
    }

    @Exclude
    public EventType getTypeAsEnum(){
        return type;
    }

    // these methods are a Firebase 9.0.0 hack to handle the enum
    public String getType(){
        if (type == null){
            return null;
        } else {
            return type.name();
        }
    }

    public void setType(String type){
        if (type == null){
            this.type = null;
        } else {
            this.type = EventType.valueOf(type);
        }
    }

    public double getCupsFood() {
        return cupsFood;
    }

    public void setCupsFood(double cupsFood) {
        this.cupsFood = cupsFood;
    }

    public boolean isNumber1() {
        return number1;
    }

    public void setNumber1(boolean number1) {
        this.number1 = number1;
    }

    public boolean isNumber2() {
        return number2;
    }

    public void setNumber2(boolean number2) {
        this.number2 = number2;
    }

    public void setWalkLength(int hours, int minutes) {
        this.walkLength = new WalkLength(hours, minutes);
    }

    public void setPrettyDate(String prettyDate) {
        this.prettyDate = prettyDate;
    }

    public String getPrettyDate() {
        if (prettyDate == null) {
            prettyDate = String.format(Locale.US, "%s %d", mContext.getResources().getStringArray(R.array.months)[getMonth() - 1], getDay());
        }
        return prettyDate;
    }

    public void setContext(Context context) {
        mContext = context;
    }

    public enum EventType {
        @SerializedName("POTTY")
        POTTY,
        @SerializedName("WALK")
        WALK,
        @SerializedName("FEED")
        FEED
    }

    public static class Builder {

        private TrackerItem item;

        public Builder() {
            item = new TrackerItem();
        }

        // ... (setters)
        public Builder setDate(String date) {
            item.setDate(date);
            return this;
        }

        public Builder setMillis(long millis) {
            item.setUtcMillis(millis);
            return this;
        }

        public Builder setId(String id) {
            item.setItemId(id);
            return this;
        }

        public Builder setPetId(String id) {
            item.setPetId(id);
            return this;
        }

        public Builder setCupsFood(double cups) {
            item.setCupsFood(cups);
            return this;
        }

        public Builder setNumber1(boolean went) {
            item.setNumber1(went);
            return this;
        }

        public Builder setNumber2(boolean went) {
            item.setNumber2(went);
            return this;
        }

        public Builder setWalkLength(int hours, int minutes) {
            item.setWalkLength(hours, minutes);
            return this;
        }

        public Builder setItemType(String type) {
            item.setItemType(type);
            return this;
        }

        public Builder setType(EventType type) {
            item.setType(type.name());
            return this;
        }

        public Builder setContext(Context context) {
            item.setContext(context);
            return this;
        }

        public TrackerItem build() {
            return item;
        }
    }
}

File: app/src/main/java/com/caitlynwiley/pettracker/models/TrackerItem.kt
package com.caitlynwiley.pettracker.models

import android.annotation.TargetApi
import android.content.Context
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Menu
import androidx.compose.material.icons.outlined.Today
import androidx.compose.ui.graphics.vector.ImageVector
import com.caitlynwiley.pettracker.R
import com.caitlynwiley.pettracker.view.icons.CustomIcons
import com.google.firebase.database.Exclude
import com.google.gson.annotations.SerializedName
import java.util.*

class TrackerItem {
    @SerializedName("date")
    var date: String? = null
        private set

    @Exclude
    @Transient
    var localTime: String? = null
        private set

    @SerializedName("item_id")
    var itemId: String = ""

    @SerializedName("item_type")
    var itemType: String = ""

    @SerializedName("month")
    var month = 0

    @SerializedName("day")
    var day = 0

    @SerializedName("year")
    var year = 0

    @SerializedName("utcmillis")
    private var utcMillis: Long = 0

    @SerializedName("event_type")
    var eventType: EventType? = null

    @SerializedName("walk_length")
    private var walkLength: WalkLength? = null

    @SerializedName("pet_id")
    var petId: String = ""

    @SerializedName("cups_food")
    var cupsFood = 0.0

    @SerializedName("number_1")
    var isNumber1 = false

    @SerializedName("number_2")
    var isNumber2 = false

    @SerializedName("pretty_date")
    private var prettyDate: String = ""

    fun setDate(c: Calendar) {
        this.date = date
        year = c[Calendar.YEAR]
        month = c[Calendar.MONTH] + 1
        day = c[Calendar.DAY_OF_MONTH]
    }

    fun setUtcMillis(utcMillis: Long) {
        this.utcMillis = utcMillis
        val c = Calendar.getInstance()
        c.timeInMillis = utcMillis
        setLocalTime(c)
        setDate(c)
    }

    @Exclude
    @TargetApi(24)
    fun setLocalTime(c: Calendar) {
        localTime = String.format(Locale.US, "%d:%02d %s",
            if (c[Calendar.HOUR] == 0) 12 else c[Calendar.HOUR],
            c[Calendar.MINUTE],
            if (c[Calendar.AM_PM] == Calendar.AM) "am" else "pm")
    }

    @get:Exclude
    val drawableResId: ImageVector
        get() {
            if (itemType.equals("day", ignoreCase = true)) return Icons.Outlined.Today
            return when (eventType) {
                EventType.POTTY -> CustomIcons.DogPoop
                EventType.FEED -> CustomIcons.DogBowl
                EventType.WALK -> CustomIcons.DogWalk
                else -> Icons.Outlined.Menu // default for now
            }
        }

    fun setWalkLength(hours: Int, minutes: Int) {
        walkLength = WalkLength(hours, minutes)
    }

    fun getDisplayDate(context: Context): String {
        if (prettyDate.isEmpty()) {
            prettyDate = String.format(Locale.US, "%s %s",
                context.resources.getStringArray(R.array.months)[month - 1],
                context.resources.getStringArray(R.array.formatted_day_of_month)[day - 1])
        }
        return prettyDate
    }

    enum class EventType {
        @SerializedName("potty") POTTY,
        @SerializedName("walk") WALK,
        @SerializedName("feed") FEED
    }

    class Builder {
        private val item: TrackerItem = TrackerItem()

        fun setMillis(millis: Long): Builder {
            item.setUtcMillis(millis)
            return this
        }

        fun setId(id: String?): Builder {
            item.itemId = id ?: ""
            return this
        }

        fun setPetId(id: String?): Builder {
            item.petId = id ?: ""
            return this
        }

        fun setCupsFood(cups: Double): Builder {
            item.cupsFood = cups
            return this
        }

        fun setNumber1(went: Boolean): Builder {
            item.isNumber1 = went
            return this
        }

        fun setNumber2(went: Boolean): Builder {
            item.isNumber2 = went
            return this
        }

        fun setWalkLength(hours: Int, minutes: Int): Builder {
            item.setWalkLength(hours, minutes)
            return this
        }

        fun setItemType(type: String): Builder {
            item.itemType = type
            return this
        }

        fun setEventType(type: EventType): Builder {
            item.eventType = type
            return this
        }

        fun build(): TrackerItem {
            return item
        }

    }
}